stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java

```diff
@@ -0,0 +1,214 @@
+        AttachedFunction onMessageFunction;
+        if (RabbitMQConstants.FUNC_ON_MESSAGE.equals(attachedFunctions[0].getName())) {
+            onMessageFunction = attachedFunctions[0];
+        } else if (RabbitMQConstants.FUNC_ON_MESSAGE.equals(attachedFunctions[1].getName())) {
+            onMessageFunction = attachedFunctions[1];
+        } else {
+            return;
+        }
+        BType[] paramTypes = onMessageFunction.paramTypes;
+        int paramSize = paramTypes.length;
+        if (paramSize > 1) {
+            dispatchMessageWithDataBinding(message, deliveryTag, onMessageFunction, properties);
+        } else {
+            dispatchMessage(message, deliveryTag, properties);
+        }
+    }
+
+    /**
+     * Dispatch messages.
+     *
+     * @param message     Message content to be dispatched to the resource function.
+     * @param deliveryTag Delivery tag of the message.
+     */
+    private void dispatchMessage(byte[] message, long deliveryTag, AMQP.BasicProperties properties) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        try {
+            Executor.submit(service, RabbitMQConstants.FUNC_ON_MESSAGE, new RabbitMQResourceCallback(countDownLatch),
+                    null, getMessageObjectValue(message, deliveryTag, properties));
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (AlreadyClosedException | BallerinaConnectorException exception) {
+            handleError(message, deliveryTag, RabbitMQConstants.DISPATCH_ERROR, properties);
+        }
+    }
+
+    /**
+     * Dispatches messages with data binding.
+     *
+     * @param message     Message content to be dispatched to the resource function.
+     * @param deliveryTag Delivery tag of the message.
+     * @param properties  Basic properties of the message.
+     */
+    private void dispatchMessageWithDataBinding(byte[] message, long deliveryTag, AttachedFunction onMessage,
+                                                AMQP.BasicProperties properties) {
+        BType[] paramTypes = onMessage.paramTypes;
+        Object[] bValues = new Object[paramTypes.length];
+        try {
+            bValues[1] = getMessageContentForType(message, paramTypes[1]);
+            bValues[0] = getMessageObjectValue(message, deliveryTag, properties);
+            CountDownLatch countDownLatch = new CountDownLatch(1);
+            Executor.submit(service, RabbitMQConstants.FUNC_ON_MESSAGE,
+                    new RabbitMQResourceCallback(countDownLatch), null, bValues);
+            countDownLatch.await();
+        } catch (BallerinaConnectorException | UnsupportedEncodingException exception) {
+            handleError(message, deliveryTag, RabbitMQConstants.DISPATCH_ERROR, properties);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
What can be done here? Can't we dispatch this to the onError as well? At least log the exception.

Reply:
handleError() function dispatches the message and the error to onError() resource function. [url](https://github.com/ballerina-platform/ballerina-lang/blob/6105407a8d2eebc78e7b92325430a613508b276f/stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java#L199)

Reply:
Have you done that in above handle InterruptedException as well? I couldn't see any dispatch to onError.

Reply:
[dispatching](https://github.com/ballerina-platform/ballerina-lang/blob/6105407a8d2eebc78e7b92325430a613508b276f/stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java#L205)

Reply:
I just want to know what happen when the InterruptedException happen? Is it going to onError or throw the exception to the upper layer to throw a panic?